runtime/src/accounts_index.rs

```diff
@@ -173,14 +173,154 @@ impl<T: 'static + Clone + IsCached> WriteAccountMapEntry<T> {
+    pub fn new(max_width: u64) -> Self {
+        assert!(max_width > 0);
+        assert!(max_width.is_power_of_two()); // power of 2 to make dividing a shift
+        let bits = BitVec::new_fill(false, max_width);
+        Self {
+            max_width,
+            bits,
+            count: 0,
+            min: 0,
+            max: 0,
+        }
+    }
+
+    fn get_address(&self, key: &u64) -> u64 {
+        Self::calc_address(self.max_width, key)
+    }
+
+    // find the array index
+    fn calc_address(max_width: u64, key: &u64) -> u64 {
+        key % max_width
+    }
+
+    fn check_range(&self, key: u64) {
+        assert!(
+            self.count == 0
+                || (self.max.saturating_sub(key) <= self.max_width as u64
+                    && key.saturating_sub(self.min) < self.max_width as u64),
+            "out of range"
+        );
+    }
+
+    pub fn insert(&mut self, key: u64) {
+        self.check_range(key);
+        let address = self.get_address(&key);
+        let value = self.bits.get(address);
+        if !value {
+            self.bits.set(address, true);
+            if self.count == 0 {
+                self.min = key;
+                self.max = key + 1;
+            } else {
+                self.min = std::cmp::min(self.min, key);
+                self.max = std::cmp::max(self.max, key + 1);
+            }
+            self.count += 1;
+        }
+    }
+
+    pub fn remove(&mut self, key: &u64) {
+        self.check_range(*key);
+        let address = self.get_address(key);
+        let value = self.bits.get(address);
+        if value {
+            self.count -= 1; // this will underflow if it tries to go below 0. That is like an assert.
+            self.bits.set(address, false);
+            self.purge(key);
+        }
+    }
+
+    // after removing 'key' where 'key' = min, make min the correct new min value
+    fn purge(&mut self, key: &u64) {
+        if key == &self.min && self.count > 0 {
+            let start = self.min;
+            for key in start..self.max {
+                if self.bits.get(key) {
+                    self.min = key;
+                    break;
+                }
+            }
+        }
+    }
+
+    pub fn contains(&self, key: &u64) -> bool {
+        let address = self.get_address(key);
+        let result = self.bits.get(address);
+        // a contains call outside min and max is allowed. The answer will be false.
+        result && (key >= &self.min && key < &self.max)
+    }
+
+    pub fn len(&self) -> usize {
+        self.count
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn clear(&mut self) {
+        let mut n = Self::new(self.max_width);
+        std::mem::swap(&mut n, self);
+    }
+
+    pub fn get_all(&self) -> Vec<u64> {
+        let mut all = Vec::with_capacity(self.count);
+        for key in self.min..self.max {
+            if self.contains(&(key as u64)) {
+                all.push(key as u64);
```

Comment:
does it need the casts? isn't key u64 here?

Also, seems like it could just get away with `self.bits.get(key)` here, it should be able to not do the range checks and the mod here.

Reply:
It does not need the casts. Hmm. I assumed clippy would fuss about that if it was happening.

Reply:
I reworked the code to get rid of the range checks (and the early exit). I think some version of the mod or 2 loops will be required to handle the case where the range wraps over the end of the bitvec size. To keep the code easy to read, I kept get_all from accessing bitvec directly.